quinn-proto/src/packet.rs

```diff
@@ -219,6 +382,65 @@ impl Header {
     }
 }
 
+pub struct PartialEncode<'a> {
+    header: &'a Header,
+    pn: Option<(usize, usize)>,
+}
+
+impl<'a> PartialEncode<'a> {
+    pub fn finish(self, buf: &mut [u8], pn_key: &PacketNumberKey, header_len: usize) {
+        let PartialEncode { header, pn } = self;
+        let payload_len = (buf.len() - header_len) as u64;
+        let (sample_offset, pn_pos, pn_len) = match header {
+            Header::Short { dst_cid, .. } => {
+                let mut sample_offset = 1 + dst_cid.len() + 4;
+                let packet_length = buf.len();
+                if sample_offset + pn_key.sample_size() > packet_length {
+                    sample_offset = packet_length - pn_key.sample_size();
+                }
+                let (pn_pos, pn_len) = pn.unwrap();
+                (sample_offset, pn_pos, pn_len)
+            }
+            Header::Initial {
+                dst_cid,
+                src_cid,
+                token,
+                ..
+            } => {
+                let sample_offset = 10
+                    + dst_cid.len()
+                    + src_cid.len()
+                    + varint::size(payload_len).unwrap()
+                    + varint::size(token.len() as u64).unwrap()
+                    + token.len();
+                let (pn_pos, pn_len) = pn.unwrap();
+                (sample_offset, pn_pos, pn_len)
```

Comment:
Probable missing bounds check. Perhaps the check/adjustment should be moved after the match?

Reply:
Which check/adjustment? The one for the short Short packets? That can really only apply to Short packets, so I don't think it generalizes.

Reply:
Why is it only applicable to short packets? It seems like long packets have exactly the same problem, with the same solution: "If the sample would extend past the end of the packet, the end of the packet is sampled." Do we have a hard lower bound of long packet payload size somewhere?

Reply:
Ah, I think I misunderstood the example code in 5.4. The example doesn't show the adjustment for the long header calculation, but from the text it becomes clear that it is still needed. That'll probably solve the Wireshark issues, then...

Reply:
Hmm, Wireshark appears to be no less confused than before. 😢